pkg/controllers/deprovisioning/orchestration/queue.go

```diff
@@ -0,0 +1,325 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package orchestration
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/avast/retry-go"
+	"github.com/mitchellh/hashstructure/v2"
+	"github.com/samber/lo"
+	"go.uber.org/multierr"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/utils/clock"
+	"knative.dev/pkg/logging"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/aws/karpenter-core/pkg/apis/v1beta1"
+	deprovisioningevents "github.com/aws/karpenter-core/pkg/controllers/deprovisioning/events"
+	"github.com/aws/karpenter-core/pkg/controllers/state"
+	"github.com/aws/karpenter-core/pkg/events"
+	"github.com/aws/karpenter-core/pkg/utils/nodeclaim"
+)
+
+// waitRetryOptions are the retry options used when trying make a successful delete call on a nodeClaim after its
+// replacements have been spun up.
+func waitRetryOptions(ctx context.Context) []retry.Option {
+	return []retry.Option{
+		retry.Context(ctx),
+		retry.Delay(2 * time.Second),
+		retry.LastErrorOnly(true),
+		retry.Attempts(10),
+		retry.MaxDelay(10 * time.Second), // 22 + (10-5)*10 =~ 1 minute in total
+	}
+}
+
+const (
+	queueBaseDelay   = 100 * time.Millisecond
+	queueMaxDelay    = 10 * time.Second
+	maxRetryDuration = 10 * time.Minute
+)
+
+type Command struct {
+	ReplacementKeys []*NodeClaimKey
+	Candidates      []*state.StateNode
+	Reason          string    `hash:"ignore"` // reason is used for metrics
+	TimeAdded       time.Time `hash:"ignore"`
+	LastError       error     `hash:"ignore"`
+}
+
+// Hash will return a uint64 representing the hash of the replacements and candidates.
+func Hash(c *Command) (uint64, error) {
+	// Hash the command so that we can connect the two underlying data structures
+	// This only needs to be done once when it's added to the queue.
+	hash, err := hashstructure.Hash(c, hashstructure.FormatV2, &hashstructure.HashOptions{
+		SlicesAsSets:    true,
+		IgnoreZeroValue: true,
+		ZeroNil:         true,
+	})
+	if err != nil {
+		return 0, fmt.Errorf("hashing command, %w", err)
+	}
+	return hash, nil
+}
+
+type NodeClaimKey struct {
+	nodeclaim.Key
+	Initialized bool // wrap a bool so we can save on recurring calls to check for initialization
```

Comment:
This feels like it _might_ be an over-optimization. It's not expensive to do a Get() call against the client (since it's calling the cache and not the actual apiserver). I don't think we expect this condition to oscillate at all so is there a reason that we feel we need to store it?

Reply:
I had also used this as a way to ensure we're only emitting initialization metrics once. I think you're probably right that we don't save much here, but I also don't think it hurts to have it, especially as we only want to emit events/metrics about initialization once.